test/absinthe/schema/sdl_render_test.exs

```diff
@@ -0,0 +1,260 @@
+defmodule SdlRenderTest do
+  use ExUnit.Case
+
+  @moduledoc """
+  https://wehavefaces.net/graphql-shorthand-notation-cheatsheet-17cd715861b6
+  https://github.com/graphql/graphql-js/blob/master/src/utilities/schemaPrinter.js
+
+  issues:
+    - schema definition order is not respected?
+
+  TODO:
+    - [ ] `Inspect` protocol for Blueprint structs!!!!!
+    - [ ] Generate `Blueprint` from introspection
+            remove from_introspection
+
+  Make tickets:
+    - default values !!
+    - deprecated reason
+    - default value: list of enums?
+  """
+
+  defmodule SdlTestSchema do
+    use Absinthe.Schema
+
+    @sdl """
+    schema {
+      query: Query
+    }
+
+    directive @foo(name: String!) on OBJECT | SCALAR
+
+    "Simple description"
+    enum Category {
+      NEWS
+      OPINION
+    }
+
+    \"""
+    A submitted post
+    Multiline description
+    \"""
+    type Post {
+      old: String
+      sweet: SweetScalar
+      "Something"
+      title: String!
+    }
+
+    type Query {
+      echo(
+        category: Category!
+        "The number of times"
+        times: Int = 10
+      ): [Category!]!
+      posts: Post
+      search(limit: Int, sort: SorterInput!): [SearchResult]
+    }
+
+    "One or the other"
+    union SearchResult = Post | User
+
+    "Sort this thing"
+    input SorterInput {
+      "By this field"
+      field: String!
+    }
+
+    scalar SweetScalar
+
+    type User {
+      name: String!
+    }
+    """
+    import_sdl @sdl
+    def sdl, do: @sdl
+  end
+
+  test "Render SDL from schema defined with SDL" do
+    {:ok, %{data: data}} = Absinthe.Schema.introspect(SdlTestSchema)
+    rendered_sdl = Absinthe.Schema.Notation.SDL.Render.from_introspection(data)
+    assert rendered_sdl == SdlTestSchema.sdl()
+  end
+
+  test "Render SDL from blueprint defined with SDL" do
+    pipeline = [
+      Absinthe.Phase.Schema.Debugger
+    ]
+
+    {:ok, blueprint, _phases} =
+      Absinthe.Pipeline.run(SdlTestSchema.__absinthe_blueprint__(), pipeline)
+
+    rendered_sdl = inspect(blueprint, pretty: true)
+
+    assert rendered_sdl == SdlTestSchema.sdl()
+  end
+
+  defmodule ClassicTestSchema do
+    use Absinthe.Schema
+
+    directive :foo do
+      arg :name, non_null(:string)
+      on [:object, :scalar]
+    end
+
+    interface :animal do
+      field :legs, non_null(:integer)
+
+      resolve_type fn
```

Comment:
@bruce How do I attach a `resolve_type` function in a SDL defined schema? Compilation fails w/o one...

Reply:
Using `hydrate/2` in the schema. In `Absinthe.Phase.Schema.Hydrate` another `apply_hydration/2` entry should be added for `:resolve_type`; here's the one we have for `:resolve`:

```elixir
  def apply_hydration(node, {:resolve, resolver}) do
    %{node | middleware: [{Absinthe.Resolution, resolver}]}
  end
```

Reply:
Note `apply_hydration/2` is a callback for the `Absinthe.Schema.Hydrator` behaviour; `Absinthe.Phase.Schema.Hydrate` implements that behaviour.